ethcore/builtin/src/lib.rs

```diff
@@ -211,89 +227,128 @@ impl Builtin {
 
 	/// Whether the builtin is activated at the given block number.
 	pub fn is_active(&self, at: u64) -> bool {
-		at >= self.activate_at
+		self.pricer.keys().rev().any(|&other_at| at >= other_at)
 	}
 }
 
 impl From<ethjson::spec::Builtin> for Builtin {
 	fn from(b: ethjson::spec::Builtin) -> Self {
-		let pricer: Box<dyn Pricer> = match b.pricing {
-			ethjson::spec::Pricing::Blake2F { gas_per_round } => {
-				Box::new(gas_per_round)
+		// TODO(niklasad1): change this to `try_from` and propogate the error accordingly
+		let native = EthereumBuiltin::from_str(&b.name).unwrap();
+		let pricer = match b.pricing {
+			ethjson::spec::builtin::Pricing::Single(pricer) => {
+				// TODO(niklasad1): change chain specs and don't enable by default!?
+				// Seems more reasonable to do it the otherway around but would be a breaking change
+				map![b.activate_at.map_or(0, Into::into) => pricer.into()]
+			}
+
+			ethjson::spec::builtin::Pricing::Multi(pricer) => {
+				assert!(!pricer.is_empty(), "chainspec contains multiple builtins; but no values found");
+
+				let mut pricers = BTreeMap::new();
+				for p in pricer {
+					pricers.insert(p.activate_at.into(), p.price.into());
+				}
+
+				pricers
+			}
+		};
+
+		Self { pricer, native }
+	}
+}
+
+
+impl From<ethjson::spec::builtin::InnerPricing> for Pricing {
+	fn from(pricing: ethjson::spec::builtin::InnerPricing) -> Self {
+		match pricing {
+			ethjson::spec::builtin::InnerPricing::Blake2F { gas_per_round } => {
+				Pricing::Blake2F(gas_per_round)
 			},
-			ethjson::spec::Pricing::Linear(linear) => {
-				Box::new(Linear {
+			ethjson::spec::builtin::InnerPricing::Linear(linear) => {
+				Pricing::Linear(Linear {
 					base: linear.base,
 					word: linear.word,
 				})
 			}
-			ethjson::spec::Pricing::Modexp(exp) => {
-				Box::new(ModexpPricer {
+			ethjson::spec::builtin::InnerPricing::Modexp(exp) => {
+				Pricing::Modexp(ModexpPricer {
 					divisor: if exp.divisor == 0 {
-						warn!(target: "builtin", "Zero modexp divisor specified. Falling back to default.");
+						warn!("Zero modexp divisor specified. Falling back to default.");
```

Comment:
```suggestion
						warn!("Zero modexp divisor specified. Falling back to default: 10.");
```

Also, why remove the `target: `? Is it not printed otherwise?

Reply:
Probably a bad rebase...

Reply:
damn, it got lost again...  needs to be fixed!